src/core.jl

```diff
@@ -165,3 +165,10 @@ function size{SA <: StaticArray}(::Type{SA})
             SMatrix{3,3}(m) # correct - size is inferrable
         """)
 end
+
+function Base.promote_rule{SA1<:StaticArray, SA2<:StaticArray}(::Type{SA1}, ::Type{SA2})
+    T = promote_type(eltype(SA1), eltype(SA2))
```

Comment:
This doesn't provide a way of defining what to do when the sizes aren't the same. Currently we get this error:

```
julia> [SVector(1,2,3), SVector(1,2)]
ERROR: The size of type `StaticArrays.SVector{S,Int64}` is not known.

If you were trying to construct (or `convert` to) a `StaticArray` you
may need to add the size explicitly as a type parameter so its size is
inferrable to the Julia compiler (or performance would be terrible). For
example, you might try

    m = zeros(3,3)
    SMatrix(m)      # this error
    SMatrix{3,3}(m) # correct - size is inferrable

 in size(::Type{StaticArrays.SVector{S,Int64}}) at /home/ferris/.julia/v0.5/StaticArrays/src/core.jl:155
 in setindex!(::Array{StaticArrays.SVector{S,Int64},1}, ::StaticArrays.SVector{3,Int64}, ::Int64) at ./array.jl:415
 in copy!(::Array{StaticArrays.SVector{S,Int64},1}, ::Tuple{StaticArrays.SVector{3,Int64},StaticArrays.SVector{2,Int64}}) at ./abstractarray.jl:479
 in vect(::StaticArrays.SVector{3,Int64}, ::Vararg{Any,N}) at ./abstractarray.jl:27
```

It turns out that this is rather awkward!! I think we might need definitions of conversion to abstract `StaticArray`s which are no-ops in the appropriate cases. Bugger.